File: push-be/internal/domain/usecase/application.go
package usecase

import (
	"context"
	"errors"
	"time"

	"github.com/google/uuid"

	"push-be/internal/domain"
)

type ApplicationService struct {
	applications domain.ApplicationStore
	jobs         domain.JobStore
	documents    domain.DocumentStore
	analyses     domain.GapAnalysisStore
	evidence     domain.EvidenceStore
	interviews   domain.InterviewStore
	submissions  domain.SubmissionStore
	approvals    *ApprovalService
	uow          domain.UnitOfWork
	adapters     map[string]bool
}

func NewApplicationService(applications domain.ApplicationStore, jobs domain.JobStore, documents domain.DocumentStore, analyses domain.GapAnalysisStore, evidence domain.EvidenceStore, interviews domain.InterviewStore, submissions domain.SubmissionStore, approvals *ApprovalService, uow domain.UnitOfWork, enabledAdapters map[string]bool) *ApplicationService {
	return &ApplicationService{
		applications: applications, jobs: jobs, documents: documents, analyses: analyses,
		evidence: evidence, interviews: interviews, submissions: submissions,
		approvals: approvals, uow: uow, adapters: enabledAdapters,
	}
}

func (s *ApplicationService) List(ctx context.Context, userID uuid.UUID, f domain.ApplicationFilter) (domain.Page[domain.Application], error) {
	limit := f.Page.EffectiveLimit()
	items, err := s.applications.List(ctx, userID, f, limit+1)
	if err != nil {
		return domain.Page[domain.Application]{}, domain.Internal()
	}
	return domain.NewPage(items, limit, func(a domain.Application) domain.Cursor {
		return domain.Cursor{CreatedAt: a.CreatedAt, ID: a.ID}
	}), nil
}

func (s *ApplicationService) Get(ctx context.Context, userID, id uuid.UUID) (*domain.Application, error) {
	a, err := s.applications.Get(ctx, userID, id)
	if err != nil {
		if errors.Is(err, domain.ErrNotFound) {
			return nil, domain.NotFound()
		}
		return nil, domain.Internal()
	}
	return a, nil
}

func (s *ApplicationService) Create(ctx context.Context, userID uuid.UUID, jobID uuid.UUID, notes string) (*domain.Application, error) {
	j, err := s.jobs.Get(ctx, userID, jobID)
	if err != nil {
		if errors.Is(err, domain.ErrNotFound) {
			return nil, domain.NotFound()
		}
		return nil, domain.Internal()
	}
	now := time.Now().UTC()
	a := &domain.Application{
		ID: uuid.New(), UserID: userID, Revision: 1,
		JobID: jobID, Company: j.Company, Title: j.Title,
		Stage: domain.StageDiscovered, Notes: notes,
		CreatedAt: now, UpdatedAt: now,
	}
	if err := s.applications.Create(ctx, a); err != nil {
		return nil, domain.Internal()
	}
	return a, nil
}

type PatchApplicationInput struct {
	ExpectedRevision int64
	Stage            *domain.ApplicationStage
	Notes            *string
	NextActionAt     *time.Time
	ClearNextAction  bool
}

func (s *ApplicationService) Patch(ctx context.Context, userID, id uuid.UUID, in PatchApplicationInput) (*domain.Application, error) {
	var out *domain.Application
	err := s.uow.Do(ctx, func(ctx context.Context) error {
		a, err := s.applications.Get(ctx, userID, id)
		if err != nil {
			if errors.Is(err, domain.ErrNotFound) {
				return domain.NotFound()
			}
			return err
		}
		if a.Revision != in.ExpectedRevision {
			return domain.RevisionConflict(a.Revision)
		}
		now := time.Now().UTC()
		if in.Stage != nil {
			to := *in.Stage
			if !domain.ValidStage(to) {
				return domain.ValidationField("stage", "unsupported stage")
			}
			if to == domain.StageApplied && a.Stage != domain.StageApplied {
				return domain.InvalidTransition("APPLIED can only be reached through submissions")
			}
			if !domain.CanPatchStage(a.Stage, to) {
				return domain.InvalidTransition("cannot move from " + string(a.Stage) + " to " + string(to))
			}
			if to != a.Stage {
				from := a.Stage
				a.Stage = to
				if to == domain.StageApplied {
					a.AppliedAt = &now
				}
				if err := s.applications.AddEvent(ctx, &domain.ApplicationEvent{
					ID: uuid.New(), UserID: userID, ApplicationID: a.ID,
					Type:      domain.EventStageChanged,
					Payload:   map[string]any{"from": string(from), "to": string(to)},
					CreatedAt: now,
				}); err != nil {
					return err
				}
			}
		}
		if in.Notes != nil {
			a.Notes = *in.Notes
		}
		if in.ClearNextAction {
			a.NextActionAt = nil
		} else if in.NextActionAt != nil {
			a.NextActionAt = in.NextActionAt
		}
		a.UpdatedAt = now
		if err := s.applications.Update(ctx, a, in.ExpectedRevision); err != nil {
			return mapRevisionErr(err)
		}
		a.Revision = in.ExpectedRevision + 1
		out = a
		return nil
	})
	if err != nil {
		return nil, err
	}
	return out, nil
}

type ImportApplicationInput struct {
	JobID     uuid.UUID
	Stage     domain.ApplicationStage
	AppliedAt time.Time
	Notes     string
	Confirmed bool
}

func (s *ApplicationService) Import(ctx context.Context, userID uuid.UUID, in ImportApplicationInput) (*domain.Application, error) {
	if !in.Confirmed {
		return nil, domain.ValidationField("confirmed", "must be true to record a past application")
	}
	if !domain.ValidStage(in.Stage) || in.Stage == domain.StageDiscovered {
		return nil, domain.ValidationField("stage", "unsupported stage for import")
	}
	j, err := s.jobs.Get(ctx, userID, in.JobID)
	if err != nil {
		if errors.Is(err, domain.ErrNotFound) {
			return nil, domain.NotFound()
		}
		return nil, domain.Internal()
	}
	now := time.Now().UTC()
	a := &domain.Application{
		ID: uuid.New(), UserID: userID, Revision: 1,
		JobID: in.JobID, Company: j.Company, Title: j.Title,
		Stage: in.Stage, Notes: in.Notes, AppliedAt: &in.AppliedAt,
		Imported: true, CreatedAt: now, UpdatedAt: now,
	}
	err = s.uow.Do(ctx, func(ctx context.Context) error {
		if err := s.applications.Create(ctx, a); err != nil {
			return err
		}
		return s.applications.AddEvent(ctx, &domain.ApplicationEvent{
			ID: uuid.New(), UserID: userID, ApplicationID: a.ID,
			Type:      domain.EventImported,
			Payload:   map[string]any{"stage": string(in.Stage), "appliedAt": in.AppliedAt},
			CreatedAt: now,
		})
	})
	if err != nil {
		return nil, domain.Internal()
	}
	return a, nil
}

func (s *ApplicationService) Timeline(ctx context.Context, userID, id uuid.UUID, page domain.PageRequest) (domain.Page[domain.ApplicationEvent], error) {
	if _, err := s.Get(ctx, userID, id); err != nil {
		return domain.Page[domain.ApplicationEvent]{}, err
	}
	limit := page.EffectiveLimit()
	items, err := s.applications.ListEvents(ctx, userID, id, page, limit+1)
	if err != nil {
		return domain.Page[domain.ApplicationEvent]{}, domain.Internal()
	}
	return domain.NewPage(items, limit, func(e domain.ApplicationEvent) domain.Cursor {
		return domain.Cursor{CreatedAt: e.CreatedAt, ID: e.ID}
	}), nil
}

type ChecklistItem struct {
	Key       string `json:"key"`
	Label     string `json:"label"`
	Completed bool   `json:"completed"`
}

type Checklist struct {
	Items             []ChecklistItem `json:"items"`
	AutomationEnabled bool            `json:"automationEnabled"`
}

func (s *ApplicationService) Checklist(ctx context.Context, userID, id uuid.UUID) (*Checklist, error) {
	a, err := s.Get(ctx, userID, id)
	if err != nil {
		return nil, err
	}
	evidenceCount, _ := s.evidence.Count(ctx, userID)
	analysisCount, _ := s.analyses.CountByApplication(ctx, userID, a.ID)
	docCount, _ := s.documents.CountFinalizedByApplication(ctx, userID, a.ID)
	interviewCount, _ := s.interviews.CountByApplication(ctx, userID, a.ID)
	return &Checklist{
		Items: []ChecklistItem{
			{Key: "evidence", Label: "Career evidence collected", Completed: evidenceCount > 0},
			{Key: "analysis", Label: "Gap analysis completed", Completed: analysisCount > 0},
			{Key: "document_finalized", Label: "Document finalized", Completed: docCount > 0},
			{Key: "applied", Label: "Application submitted", Completed: a.AppliedAt != nil || domain.TerminalStage(a.Stage) || a.Stage == domain.StageApplied || stageAfterApplied(a.Stage)},
			{Key: "interview_prep", Label: "Interview session scheduled", Completed: interviewCount > 0},
		},
		AutomationEnabled: false,
	}, nil
}

func stageAfterApplied(s domain.ApplicationStage) bool {
	switch s {
	case domain.StageScreening, domain.StageInterview, domain.StageOffer:
		return true
	}
	return false
}

type CreateDraftInput struct {
	ExpectedRevision   int64
	Mode               domain.SubmissionMode
	Adapter            *string
	DocumentVersionIDs []uuid.UUID
	ConfirmedSubmitted bool
}

func (s *ApplicationService) CreateDraft(ctx context.Context, userID, appID uuid.UUID, in CreateDraftInput) (*domain.SubmissionDraft, error) {
	var out *domain.SubmissionDraft
	err := s.uow.Do(ctx, func(ctx context.Context) error {
		a, err := s.applications.Get(ctx, userID, appID)
		if err != nil {
			if errors.Is(err, domain.ErrNotFound) {
				return domain.NotFound()
			}
			return err
		}
		if a.Revision != in.ExpectedRevision {
			return domain.RevisionConflict(a.Revision)
		}
		if in.Mode != domain.SubmissionManual && in.Mode != domain.SubmissionAdapter {
			return domain.ValidationField("mode", "must be MANUAL_RECORD or ADAPTER")
		}
		if in.Mode == domain.SubmissionAdapter {
			if in.Adapter == nil || !s.adapters[*in.Adapter] {
				return domain.FeatureDisabled("adapter submission is not enabled for this site")
			}
		} else if in.Adapter != nil {
			return domain.ValidationField("adapter", "adapter is only valid with ADAPTER mode")
		}
		if len(in.DocumentVersionIDs) == 0 {
			return domain.ValidationField("documentVersionIds", "at least one document version is required")
		}
		for _, vid := range in.DocumentVersionIDs {
			v, err := s.documents.GetVersion(ctx, userID, uuid.Nil, vid)
			if err != nil {
				return domain.ValidationField("documentVersionIds", "version "+vid.String()+" not found")
			}
			if v.ApplicationID != appID {
				return domain.ValidationField("documentVersionIds", "version "+vid.String()+" belongs to another application")
			}
			doc, err := s.documents.Get(ctx, userID, v.DocumentID)
			if err != nil {
				return domain.Internal()
			}
			if doc.FinalizedVersionID == nil || *doc.FinalizedVersionID != vid {
				return domain.DocumentNotFinalized("document version " + vid.String() + " is not finalized")
			}
		}
		d := &domain.SubmissionDraft{
			ID: uuid.New(), UserID: userID, ApplicationID: appID,
			ApplicationRevision: a.Revision, Mode: in.Mode, Adapter: in.Adapter,
			DocumentVersionIDs: in.DocumentVersionIDs, ConfirmedSubmitted: in.ConfirmedSubmitted,
			PayloadHash: domain.HashJSON(map[string]any{
				"applicationId": appID, "applicationRevision": a.Revision,
				"mode": in.Mode, "adapter": in.Adapter, "documentVersionIds": in.DocumentVersionIDs,
			}),
			CreatedAt: time.Now().UTC(),
		}
		if err := s.submissions.CreateDraft(ctx, d); err != nil {
			return err
		}
		out = d
		return nil
	})
	if err != nil {
		return nil, err
	}
	return out, nil
}

type SubmitInput struct {
	ExpectedRevision int64
	DraftID          uuid.UUID
	ApprovalID       uuid.UUID
}

func (s *ApplicationService) Submit(ctx context.Context, userID, appID uuid.UUID, in SubmitInput) (*domain.Submission, error) {
	var out *domain.Submission
	err := s.uow.Do(ctx, func(ctx context.Context) error {
		a, err := s.applications.Get(ctx, userID, appID)
		if err != nil {
			if errors.Is(err, domain.ErrNotFound) {
				return domain.NotFound()
			}
			return err
		}
		if a.Revision != in.ExpectedRevision {
			return domain.RevisionConflict(a.Revision)
		}
		d, err := s.submissions.GetDraft(ctx, userID, in.DraftID)
		if err != nil {
			return domain.ValidationField("draftId", "draft not found")
		}
		if d.ApplicationID != appID {
			return domain.ValidationField("draftId", "draft belongs to another application")
		}
		if d.ApplicationRevision != a.Revision {
			return domain.ApprovalStale("application changed since the draft was created; create a new draft and approval")
		}
		if _, err := s.approvals.Consume(ctx, userID, in.ApprovalID, domain.ApprovalApplicationSubmit, appID, d.ID, d.PayloadHash); err != nil {
			return err
		}
		if d.Mode == domain.SubmissionManual && !d.ConfirmedSubmitted {
			return domain.ValidationField("confirmedSubmitted", "manual submission requires confirming the external submission completed")
		}
		now := time.Now().UTC()
		sub := &domain.Submission{
			ID: uuid.New(), UserID: userID, ApplicationID: appID,
			Mode: d.Mode, Adapter: d.Adapter, Status: domain.SubmissionSucceeded,
			DocumentVersionIDs: d.DocumentVersionIDs, ApprovalID: in.ApprovalID,
			CreatedAt: now, UpdatedAt: now,
		}
		if err := s.submissions.Create(ctx, sub); err != nil {
			return err
		}
		if domain.TerminalStage(a.Stage) {
			return domain.InvalidTransition("application is closed")
		}
		if !domain.TerminalStage(a.Stage) && a.Stage != domain.StageApplied && !stageAfterApplied(a.Stage) {
			from := a.Stage
			a.Stage = domain.StageApplied
			a.AppliedAt = &now
			a.UpdatedAt = now
			if err := s.applications.Update(ctx, a, in.ExpectedRevision); err != nil {
				return mapRevisionErr(err)
			}
			a.Revision = in.ExpectedRevision + 1
			if err := s.applications.AddEvent(ctx, &domain.ApplicationEvent{
				ID: uuid.New(), UserID: userID, ApplicationID: a.ID,
				Type:      domain.EventStageChanged,
				Payload:   map[string]any{"from": string(from), "to": string(domain.StageApplied)},
				CreatedAt: now,
			}); err != nil {
				return err
			}
		}
		if err := s.applications.AddEvent(ctx, &domain.ApplicationEvent{
			ID: uuid.New(), UserID: userID, ApplicationID: a.ID,
			Type:      domain.EventSubmission,
			Payload:   map[string]any{"submissionId": sub.ID, "mode": d.Mode, "status": sub.Status},
			CreatedAt: now,
		}); err != nil {
			return err
		}
		out = sub
		return nil
	})
	if err != nil {
		return nil, err
	}
	return out, nil
}

func (s *ApplicationService) ListSubmissions(ctx context.Context, userID, appID uuid.UUID, page domain.PageRequest) (domain.Page[domain.Submission], error) {
	if _, err := s.Get(ctx, userID, appID); err != nil {
		return domain.Page[domain.Submission]{}, err
	}
	limit := page.EffectiveLimit()
	items, err := s.submissions.List(ctx, userID, appID, page, limit+1)
	if err != nil {
		return domain.Page[domain.Submission]{}, domain.Internal()
	}
	return domain.NewPage(items, limit, func(s domain.Submission) domain.Cursor {
		return domain.Cursor{CreatedAt: s.CreatedAt, ID: s.ID}
	}), nil
}

File: push-be/app/domain/services/application.py
from __future__ import annotations
from datetime import datetime, timezone
from typing import Optional
from uuid import UUID, uuid4

from app.db import DB, NotFoundError
from app.domain import entities as ent
from app.domain.errors import (
    document_not_finalized, feature_disabled, internal, invalid_transition,
    map_revision_err, not_found, revision_conflict, validation_field,
    approval_stale,
)
from app.domain.services.approval import ApprovalService, hash_json
from app.infrastructure.store_applications import ApplicationStore, SubmissionStore
from app.infrastructure.store_calendar import InterviewStore
from app.infrastructure.store_documents import DocumentStore
from app.infrastructure.store_evidence import EvidenceStore
from app.infrastructure.store_jobs import GapAnalysisStore, JobStore

SUBMISSION_MANUAL = "MANUAL_RECORD"
SUBMISSION_ADAPTER = "ADAPTER"
SUBMISSION_SUCCEEDED = "SUCCEEDED"


def _now() -> datetime:
    return datetime.now(timezone.utc)


class ApplicationService:
    def __init__(self, db: DB, approvals: ApprovalService, adapters: dict):
        self.db = db
        self.applications = ApplicationStore(db)
        self.jobs = JobStore(db)
        self.documents = DocumentStore(db)
        self.analyses = GapAnalysisStore(db)
        self.evidence = EvidenceStore(db)
        self.interviews = InterviewStore(db)
        self.submissions = SubmissionStore(db)
        self.approvals = approvals
        self.adapters = adapters or {}

    async def list(self, user_id: UUID, stage: Optional[str], query: str, page):
        try:
            return await self.applications.list(user_id, stage, query, page)
        except Exception:
            raise internal()

    async def get(self, user_id: UUID, id_: UUID) -> ent.Application:
        try:
            return await self.applications.get(user_id, id_)
        except NotFoundError:
            raise not_found()
        except Exception:
            raise internal()

    async def create(self, user_id: UUID, job_id: UUID, notes: str) -> ent.Application:
        try:
            j = await self.jobs.get(user_id, job_id)
        except NotFoundError:
            raise not_found()
        except Exception:
            raise internal()
        now = _now()
        a = ent.Application(
            id=uuid4(), user_id=user_id, revision=1, job_id=job_id, company=j.company,
            title=j.title, stage="DISCOVERED", notes=notes, created_at=now, updated_at=now)
        try:
            await self.applications.create(a)
        except Exception:
            raise internal()
        return a

    async def patch(self, user_id: UUID, id_: UUID, expected: int,
                    stage: Optional[str], notes: Optional[str],
                    next_action_at, clear_next_action: bool) -> ent.Application:
        out = None

        async def work():
            nonlocal out
            try:
                a = await self.applications.get(user_id, id_)
            except NotFoundError:
                raise not_found()
            if a.revision != expected:
                raise revision_conflict(a.revision)
            now = _now()
            if stage is not None:
                to = stage
                if not ent.valid_stage(to):
                    raise validation_field("stage", "unsupported stage")
                if to == ent.STAGE_APPLIED and a.stage != ent.STAGE_APPLIED:
                    raise invalid_transition(
                        "APPLIED can only be reached through submissions")
                if not ent.can_patch_stage(a.stage, to):
                    raise invalid_transition(
                        f"cannot move from {a.stage} to {to}")
                if to != a.stage:
                    from_ = a.stage
                    a.stage = to
                    if to == ent.STAGE_APPLIED:
                        a.applied_at = now
                    await self.applications.add_event(ent.ApplicationEvent(
                        id=uuid4(), user_id=user_id, application_id=a.id,
                        type=ent.EVENT_STAGE_CHANGED,
                        payload={"from": from_, "to": to}, created_at=now))
            if notes is not None:
                a.notes = notes
            if clear_next_action:
                a.next_action_at = None
            elif next_action_at is not None:
                a.next_action_at = next_action_at
            a.updated_at = now
            try:
                await self.applications.update(a, expected)
            except Exception as err:
                raise map_revision_err(err)
            a.revision = expected + 1
            out = a

        await self.db.do(work)
        return out

    async def import_(self, user_id: UUID, job_id: UUID, stage: str,
                      applied_at: datetime, notes: str,
                      confirmed: bool) -> ent.Application:
        if not confirmed:
            raise validation_field(
                "confirmed", "must be true to record a past application")
        if not ent.valid_stage(stage) or stage == "DISCOVERED":
            raise validation_field("stage", "unsupported stage for import")
        try:
            j = await self.jobs.get(user_id, job_id)
        except NotFoundError:
            raise not_found()
        except Exception:
            raise internal()
        now = _now()
        a = ent.Application(
            id=uuid4(), user_id=user_id, revision=1, job_id=job_id, company=j.company,
            title=j.title, stage=stage, notes=notes, applied_at=applied_at,
            imported=True, created_at=now, updated_at=now)

        async def work():
            await self.applications.create(a)
            await self.applications.add_event(ent.ApplicationEvent(
                id=uuid4(), user_id=user_id, application_id=a.id,
                type=ent.EVENT_IMPORTED,
                payload={"stage": stage, "appliedAt": applied_at}, created_at=now))

        try:
            await self.db.do(work)
        except Exception:
            raise internal()
        return a

    async def timeline(self, user_id: UUID, id_: UUID, page):
        await self.get(user_id, id_)
        try:
            return await self.applications.list_events(user_id, id_, page)
        except Exception:
            raise internal()

    async def checklist(self, user_id: UUID, id_: UUID) -> dict:
        a = await self.get(user_id, id_)
        evidence_count = await self.evidence.count(user_id)
        analysis_count = await self.analyses.count_by_application(user_id, a.id)
        doc_count = await self.documents.count_finalized_by_application(user_id, a.id)
        interview_count = await self.interviews.count_by_application(user_id, a.id)
        applied = (a.applied_at is not None or ent.terminal_stage(a.stage)
                   or a.stage == ent.STAGE_APPLIED or ent.stage_after_applied(a.stage))
        return {
            "items": [
                {"key": "evidence", "label": "Career evidence collected",
                 "completed": evidence_count > 0},
                {"key": "analysis", "label": "Gap analysis completed",
                 "completed": analysis_count > 0},
                {"key": "document_finalized", "label": "Document finalized",
                 "completed": doc_count > 0},
                {"key": "applied", "label": "Application submitted",
                 "completed": applied},
                {"key": "interview_prep", "label": "Interview session scheduled",
                 "completed": interview_count > 0},
            ],
            "automationEnabled": False,
        }

    async def create_draft(self, user_id: UUID, app_id: UUID, expected: int,
                           mode: str, adapter: Optional[str],
                           document_version_ids: list[UUID],
                           confirmed_submitted: bool) -> ent.SubmissionDraft:
        out = None

        async def work():
            nonlocal out
            try:
                a = await self.applications.get(user_id, app_id)
            except NotFoundError:
                raise not_found()
            if a.revision != expected:
                raise revision_conflict(a.revision)
            if mode not in (SUBMISSION_MANUAL, SUBMISSION_ADAPTER):
                raise validation_field("mode", "must be MANUAL_RECORD or ADAPTER")
            if mode == SUBMISSION_ADAPTER:
                if not adapter or not self.adapters.get(adapter):
                    raise feature_disabled(
                        "adapter submission is not enabled for this site")
            elif adapter is not None:
                raise validation_field(
                    "adapter", "adapter is only valid with ADAPTER mode")
            if not document_version_ids:
                raise validation_field(
                    "documentVersionIds", "at least one document version is required")
            nil = UUID(int=0)
            for vid in document_version_ids:
                try:
                    v = await self.documents.get_version(user_id, nil, vid)
                except Exception:
                    raise validation_field(
                        "documentVersionIds", "version " + str(vid) + " not found")
                if v.application_id != app_id:
                    raise validation_field(
                        "documentVersionIds",
                        "version " + str(vid) + " belongs to another application")
                try:
                    doc = await self.documents.get(user_id, v.document_id)
                except Exception:
                    raise internal()
                if doc.finalized_version_id is None or doc.finalized_version_id != vid:
                    raise document_not_finalized(
                        "document version " + str(vid) + " is not finalized")
            d = ent.SubmissionDraft(
                id=uuid4(), user_id=user_id, application_id=app_id,
                application_revision=a.revision, mode=mode, adapter=adapter,
                document_version_ids=document_version_ids,
                confirmed_submitted=confirmed_submitted,
                payload_hash=hash_json({
                    "applicationId": app_id, "applicationRevision": a.revision,
                    "mode": mode, "adapter": adapter,
                    "documentVersionIds": document_version_ids}),
                created_at=_now())
            await self.submissions.create_draft(d)
            out = d

        await self.db.do(work)
        return out

    async def submit(self, user_id: UUID, app_id: UUID, expected: int,
                     draft_id: UUID, approval_id: UUID) -> ent.Submission:
        out = None

        async def work():
            nonlocal out
            try:
                a = await self.applications.get(user_id, app_id)
            except NotFoundError:
                raise not_found()
            if a.revision != expected:
                raise revision_conflict(a.revision)
            try:
                d = await self.submissions.get_draft(user_id, draft_id)
            except Exception:
                raise validation_field("draftId", "draft not found")
            if d.application_id != app_id:
                raise validation_field(
                    "draftId", "draft belongs to another application")
            if d.application_revision != a.revision:
                raise approval_stale(
                    "application changed since the draft was created; "
                    "create a new draft and approval")
            await self.approvals.consume(
                user_id, approval_id, ent.APPROVAL_APPLICATION_SUBMIT,
                app_id, d.id, d.payload_hash)
            if d.mode == SUBMISSION_MANUAL and not d.confirmed_submitted:
                raise validation_field(
                    "confirmedSubmitted",
                    "manual submission requires confirming the external submission completed")
            now = _now()
            sub = ent.Submission(
                id=uuid4(), user_id=user_id, application_id=app_id,
                mode=d.mode, adapter=d.adapter, status=SUBMISSION_SUCCEEDED,
                document_version_ids=d.document_version_ids, approval_id=approval_id,
                created_at=now, updated_at=now)
            await self.submissions.create(sub)
            if ent.terminal_stage(a.stage):
                raise invalid_transition("application is closed")
            if (a.stage != ent.STAGE_APPLIED
                    and not ent.stage_after_applied(a.stage)):
                from_ = a.stage
                a.stage = ent.STAGE_APPLIED
                a.applied_at = now
                a.updated_at = now
                try:
                    await self.applications.update(a, expected)
                except Exception as err:
                    raise map_revision_err(err)
                a.revision = expected + 1
                await self.applications.add_event(ent.ApplicationEvent(
                    id=uuid4(), user_id=user_id, application_id=a.id,
                    type=ent.EVENT_STAGE_CHANGED,
                    payload={"from": from_, "to": ent.STAGE_APPLIED}, created_at=now))
            await self.applications.add_event(ent.ApplicationEvent(
                id=uuid4(), user_id=user_id, application_id=a.id,
                type=ent.EVENT_SUBMISSION,
                payload={"submissionId": sub.id, "mode": d.mode, "status": sub.status},
                created_at=now))
            out = sub

        await self.db.do(work)
        return out

    async def list_submissions(self, user_id: UUID, app_id: UUID, page):
        await self.get(user_id, app_id)
        try:
            return await self.submissions.list(user_id, app_id, page)
        except Exception:
            raise internal()
